developer/src/kmc-keyboard/src/kvk/kvks-file-reader.ts

```diff
@@ -0,0 +1,167 @@
+import * as xml2js from 'xml2js';
+import KVKSourceFile from './kvks-file.js';
+import CompilerCallbacks from '../compiler/callbacks.js';
+import Ajv from 'ajv';
+import { CompilerMessages } from '../compiler/messages.js';
+import { boxXmlArray } from '../util/util.js';
+import { VisualKeyboard, VisualKeyboardHeaderFlags, VisualKeyboardKey, VisualKeyboardKeyFlags, VisualKeyboardLegalShiftStates, VisualKeyboardShiftState } from './visual-keyboard.js';
+import { USVirtualKeyCodes } from '../ldml-keyboard/virtual-key-constants.js';
+import { BUILDER_KVK_HEADER_VERSION } from './kvk-file.js';
+
+export default class KVKSFileReader {
+  private readonly callbacks: CompilerCallbacks;
+
+  constructor (callbacks: CompilerCallbacks) {
+    this.callbacks = callbacks;
+  }
+
+  public read(file: Uint8Array): VisualKeyboard {
+    let source = this.internalRead(file);
+    return this.transform(source);
+  }
+
+  public internalRead(file: Uint8Array): KVKSourceFile {
+    let source: KVKSourceFile;
+
+    const parser = new xml2js.Parser({
+      explicitArray: false,
+      mergeAttrs: true,
+      includeWhiteChars: true,
+      normalize: false,
+      emptyTag: {} as any
+      // Why "as any"? xml2js is broken:
+      // https://github.com/Leonidas-from-XIV/node-xml2js/issues/648 means
+      // that an old version of `emptyTag` is used which doesn't support
+      // functions, but DefinitelyTyped is requiring use of function or a
+      // string. See also notes at
+      // https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59259#issuecomment-1254405470
+      // An alternative fix would be to pull xml2js directly from github
+      // rather than using the version tagged on npmjs.com.
```

Comment:
i assume this was before the bugfix you found also

Reply:
yes.